@elpaso I'm a bit lost with the saveConnection / deleteConnection API and in particular with the tests performed by tests/src/python/test_qgsproviderconnection_base.py which keep failing for me.

There are 2 tests failing.

First one:

  - line 261, in _test_operations:
```
    self.assertEqual(list(md.connections().values()), [])
AssertionError: Lists differ: [<qgis._core.QgsAbstractDatabaseProviderConnection object at 0x7f9504f683a8>] != []
```

I'm guessing this is due to me creating that new connection and not cleaning up afterwards ?
How is a connection removed ? I see calls to `md.deleteConnection` but they are passed a connection _name_ rather than a connection object, so how do I tell what to pass there ?
I've also tried to saveConnection(conn2, 'newname') and deleteConnection('newname') but that test keeps failing.

--- a/src/providers/postgres/qgspostgresconn.cpp
+++ b/src/providers/postgres/qgspostgresconn.cpp
@@ -1074,7 +1074,23 @@ QString QgsPostgresConn::postgisVersion()
   mTopologyAvailable = false;
   if ( mPostgisVersionMajor > 1 )
   {
-    QgsPostgresResult result( PQexec( QStringLiteral( "SELECT EXISTS ( SELECT c.oid FROM pg_class AS c JOIN pg_namespace AS n ON c.relnamespace=n.oid WHERE n.nspname='topology' AND c.relname='topology' )" ) ) );
+    // NOTE: CASE syntax is used to avoid an exception when
+    //       topology.topology does not exist
+    // See
+    // https://www.postgresql.org/docs/current/sql-expressions.html#SYNTAX-EXPRESS-EVAL
+    QgsPostgresResult result(
+      PQexec(
+        QStringLiteral(
+          "SELECT has_schema_privilege(n.oid, 'usage')"
+          "   AND has_table_privilege(t.oid, 'select')"
+          "   AND has_table_privilege(l.oid, 'select')"
+          "  FROM pg_namespace n, pg_class t, pg_class l"
+          " WHERE n.nspname = 'topology'"
+          "   AND t.relnamespace = n.oid"
+          "   AND l.relnamespace = n.oid"
+          "   AND t.relname = 'topology'"
+          "   AND l.relname = 'layer'"
+        ) ) );
     if ( result.PQntuples() >= 1 && result.PQgetvalue( 0, 0 ) == QLatin1String( "t" ) )
     {
       mTopologyAvailable = true;

--- a/src/providers/postgres/qgspostgresconn.cpp
+++ b/src/providers/postgres/qgspostgresconn.cpp
@@ -353,10 +353,9 @@ QgsPostgresConn::QgsPostgresConn( const QString &conninfo, bool readOnly, bool s
     {
       QgsMessageLog::logMessage( tr( "Your PostGIS installation has no GEOS support. Feature selection and identification will not work properly. Please install PostGIS with GEOS support (http://geos.refractions.net)" ), tr( "PostGIS" ) );
     }
-
-    if ( hasTopology() )
+    else
     {
-      QgsDebugMsg( QStringLiteral( "Topology support available!" ) );
+      QgsDebugMsg( QStringLiteral( "GEOS support available!" ) );
     }
   }
 
@@ -1095,6 +1094,15 @@ QString QgsPostgresConn::postgisVersion()
     {
       mTopologyAvailable = true;
     }
+  }
+
+  if ( mTopologyAvailable )
+  {
+    QgsDebugMsg( QStringLiteral( "Topology support available!" ) );
+  }
+  else
+  {
+    QgsDebugMsg( QStringLiteral( "Topology support not available :(" ) );
   }
 
   mGotPostgisVersion = true;

--- a/src/providers/postgres/qgspostgresconn.cpp
+++ b/src/providers/postgres/qgspostgresconn.cpp
@@ -408,8 +408,11 @@ QStringList QgsPostgresConn::supportedSpatialTypes() const
   supportedSpatialTypes << quotedValue( "geometry" )
                         << quotedValue( "geography" )
                         << quotedValue( "pcpatch" );
-  if ( hasRaster() ) supportedSpatialTypes << quotedValue( "raster" );
-  if ( hasTopology() ) supportedSpatialTypes << quotedValue( "topogeometry" );
+  if ( hasRaster() )
+    supportedSpatialTypes << quotedValue( "raster" );
+
+  if ( hasTopology() )
+    supportedSpatialTypes << quotedValue( "topogeometry" );
 
   return supportedSpatialTypes;
 }
@@ -1081,7 +1084,7 @@ QString QgsPostgresConn::postgisVersion() const
 
   if ( mTopologyAvailable )
   {
-    QgsDebugMsg( QStringLiteral( "Topology support available!" ) );
+    QgsDebugMsg( QStringLiteral( "Topology support available :)" ) );
   }
   else
   {

